model/src/app_data.rs

```diff
@@ -0,0 +1,88 @@
+//! Contains the app_data file structures to define additional data about tx origin
+
+use crate::h160_hexadecimal::{self};
+use anyhow::{anyhow, Result};
+use cid::multihash::{Code, MultihashDigest};
+use primitive_types::{H160, H256};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use std::convert::TryInto;
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct MetaData {
+    #[serde(with = "h160_hexadecimal")]
+    pub referrer: H160,
+}
+
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Debug, Deserialize, Serialize, Hash)]
+#[serde(rename_all = "camelCase")]
+pub struct AppData {
+    pub version: String,
+    pub app_code: String,
+    pub meta_data: MetaData,
+}
+impl AppData {
+    pub fn sha_hash(&self) -> Result<H256> {
+        let string = serde_json::to_string(self)?;
```

Comment:
yes, struct -> string -> sha is always deterministic, as the struct is unique.

but also the critical part: Json(string) -> struct is deterministic

Reply:
I understand that it is deterministic in practice but I'm not sure if the serialization libraries guarantee this. Maybe a serde update would reorder the fields or otherwise change the serialization in a way that doesn't affect the semantics of the json.
And I guess this always has to match the frontend too? Maybe the spec should say exactly how the hash is calculated? For example we could say that instead of hashing the json string we create a new string that is the concatenation of version + app_code + referrer_as_hex or something like that.

Reply:
>but I'm not sure if the serialization libraries guarantee this.
even if they don't guarantee it, I think that is quite unlikely that something will change. If worst comes to worst, we don't have to update the lib. 

But yeah, we could also go your other route. It just has one disadvantage, with each new version and new struct fields, we would have to change how it's being calculated.

wdyt @alfetopito ?

Reply:
I would prefer to have a deterministic json serialization.

The suggestion to hash a subset of fields is limited to the fields we currently have, but we expect to have more in the future. It' would be a uneccessary maintenance burden IMO.

Reply:
The question is what is deterministic json serialization. Afaik there is no such generally accepted standard and there are several ways you could implement the idea. The problem is that the library the frontend is using might not be doing the same thing as the one the backend is using even though both are trying to be deterministic. We would have to be sure that the implementations match but that probably involves implementing a whole json serializer. This is why I think it is easier to manually specify how the hashing works.

Reply:
>We would have to be sure that the implementations match but that probably involves implementing a whole json serializer. 

I think that it is okay if we rely on API calls for that and have it only implemented in the backend. In the other pending PRs, the front-end can send any json and the backend would store it and send back its hash

Reply:
I'm not a fan of this approach. The way this hash is calculated feels like something that should be part of the spec and be "implementation defined". It should be possible without going through a running driver. And even that we'd still need to ensure that this doesn't change which serde_json doesn't guarantee. Alternatively there might be an approach that doesn't need this kind of hash? I'd have to think about it more.

Reply:
>Alternatively there might be an approach that doesn't need this kind of hash?

We need to compress many data into one variable that can be put into appData of an order. I think we need some compression like sha hashes...

Reply:
The property ordering discussion is interesting!

Given that the API is just a temporal workaround until we can store this information on IPFS, we should maybe have it implement a similar interface as IPFS? 
That is you can "store" a blob (string) on the db and it will return you the key (IPFS hash) of what you stored. It's then up to the frontend to put that same hash on the order data.

Reply:
I had a long call with @alfetopito and anna, and we concluded:

- we will try to minimize complexity now, as this is not an essential component
- for easy storage and easy queries, we will store most of the data just as a string blob, though we will keep the appHash, appCode, and referrer as separate fields to enable building easy queries for later.
- The appDataHash will just be calculated as the hash of the string blob stored. Even if this is results in different hashes for identical data - e.g. just the same values, but different order - we don't see a problem with that. Our queries and referral logic should still work.

Wdyt? Is there something we are missing?

Reply:
I like that idea. You could even reject the request to store the blob if it doesn deserialize into your well defined struct that we want to use. I think the main concern here was that the ordering of fields in the string shouldn't decide whether or not the blob is valid (as different encoding libraries might change that ordering).

Reply:
My initial thinking is, frontend doesn't even need the API to attach the information. The only reason we are doing it is for "availability" reasons, to make sure that IPFS will always have this info.
Another reason to use the API is to make it also easier for Balancer.

However, strictly speaking, we don't need an API for just storing the IPFS, this can be indexed by the API pulling from the order. Since the API can also get the IPFS document easily and this is a simple JSON is easy to index it. 
Using a service like Pinata we can make sure the availability problem goes away and we would be more decentralized in info fetching. 

The issue with the code for me, is that you rely on generating the document yourself, which I think is constraining for these reasons: 
- Now you are the one defining the format of the document. If a client wants to add additional information, they can't. Think of a client wishing to add additional metadata that the API is still not aware, they won't be able to do that. The document was created in a flexible fashion, so we can keep indexers (like the API) backward compatible. New metadatas are just ignored. Futures versions of the API (or indexers can start to handle metadatas retroactivelly)
- Also, here you are the one defining the structure of the JSON, you will run into issues when you evolve this data format because if you "reprocess" old documents you will get different hash unless you start to add some workarounds and versioning things (I see there's no version in the metadata, so you might even have issues doing this)
- On top of this, if you don't upload the document to IPFS other indexers, like TheGraph, or any other will run into issues to get the metadata of the order

I'm happy either way, if we:
- Use the API, but we store the original file or the blob, and we hash the blob/file blindly, not a DTO on the API definition. If we go this path, I would consider a requirement that this hashing yields the same hash as if I would upload it to IPFS, so we make sure in the future we can upload all
- Not use the API for uploading the IPFS link, for that we can use a service like Pinata, and then the API only needs to index the hash from the signed order (same as all other indexers)


For example, for this document here:
```ts
{
  "version": "1.0.0",
  "appCode": "CowSwap",
  "metadata": {
    "referrer": {
      "version": "1.2.3",
      "kind": "Referrer",
      "referrer": "0x424a46612794dbb8000194937834250Dc723fFa5"
    }
  }
}
```

we would get `QmT8qCMvCtXtWpMeNwk7AeHHUnJNfv83DxDawLMEpMz694` https://gateway.pinata.cloud/ipfs/QmT8qCMvCtXtWpMeNwk7AeHHUnJNfv83DxDawLMEpMz694

Reply:
Agree, that the API should not limit the format that can be uploaded but expect arbitrary strings. It may reject the string if it cannot extract the relevant information but should allow additional info.

The nice thing about uploading it to Pinata would be that aggregation of the referral information could theoretically be done in a completely separate service (e.g. based on the Graph trade data or trade event indexing and a IPFS client). But given we wanted this quick the easiest way seemed to do it in our API database directly.

Reply:
For me is fine to not upload to IPFS if we can upload it later retroactively. The point I tried to raise was that if we don't, at least we should make sure we hash the original JSON (as long as it has the right format and can extract at least the AppCode). This way, in the future when we implement the Graph, we can also upload all the old document so TheGraph can consume them. 

> The nice thing about uploading it to Pinata would be that aggregation of the referral information could theoretically be done in a completely separate service (e.g. based on the Graph trade data or trade event indexing and a IPFS client). But given we wanted this quick the easiest way seemed to do it in our API database directly.

Fine with me, we will also need to do a FK in the trades, right? 

Regarding the model, I think we will need a table with a mapping of user accounts to their referral. This is because, I believe that if we implement some affiliate logics in the backend we need to then:
- detect when is the first time a user trade
- If the trade is for a user that never traded before. Does it have referal information? If it does, we record it. Note that it shouldn't be relevant and we shouldn't record the referal associated to an user when the user traded before. It's meant to be an incentive only for new users.

Reply:
I think we can use EIP-712 structured data hashing for generating deterministic hashes for structured data. This way we have an algorithm for generating the hash based on the data itself and not based on some canonical representation.

Additionally, we already support this kind of hashing for signing in both the code bases.

Reply:
@nlordell how would that work? We would first derive the hashStruct from the json object and then calculate the EIP-712 hash based on this struct? This would also result in different sha_hashes, if the hashes of two different permutations of app_data fields is calculated, right? 
But it solves the issues with the fact that just an additional empty_space already modifies the hash.